Data/Graph/Inductive/Query/SP/DAG.hs

```diff
@@ -0,0 +1,200 @@
+
+-- | Returns the longest possible path in the graph.
+longestPath :: (Graph gr, Real b, Foldable f)
+    => Int  -- ^ length of topsort
+    -> f Node -- ^ topsort
+    -> gr a b
+    -> IO Path
+longestPath l t g = do
+    (dist,predecessor) <- runPaths Nothing (Number 0) (Number 0) (<) l t g
+    return $ tracePath (maxIndex dist) predecessor
+
+-- | Traces a path backwards through a vector of predecessors,
+--   where a vector of predecessors is a vector such that the value 
+--   in the vector at any position n where n is a node in the graph 
+--   is the predecessor to that node.
+tracePath :: MonadPlus m 
+    => Node -- ^ End node of path
+    -> Vector (Maybe Node) -- ^ Vector of predecessors
+    -> m Node -- ^ MonadPlus of Path
+tracePath n v = 
+    case v ! n of
+         Just node 
+              -> mplus (return n) $ tracePath node v
+         Nothing
+              -> mplus (return n) mzero
+
+-- | A reasonably generalized function used for generating 
+--   a vector of distances and predecessors obtained for a given
+--   comparison operation, an optional source node, and so on.
+runPaths :: (Graph gr, Real b, Foldable f)
+    => Maybe Node -- Source node
+    -> Number b -- ^ Source node initialization value
+    -> Number b -- ^ Distances vector init values
+    -> (Number b -> Number b -> Bool) -- ^ Weight multiplie
+    -> Int -- ^ Length of the top sort
+    -> f Node -- ^ Topological sorting of the graph in some foldable format
+    -> gr a b -- ^ A graph.
+-- | A vector of distances and predecessors. The distances represent maximum or
+--   minimum distance to a given node, and the predecessors represent the node that
+--   linked to that node on the maximum or minimum path.
+    -> IO (Vector (Number b), Vector (Maybe Node))
+runPaths s v ds c l t g = 
+    freezeVecs =<< runPaths' s v ds c l t g
+
+freezeVecs ::
+    (IOVector a, IOVector b)
+    -> IO (Vector a, Vector b)
+freezeVecs (v1,v2) = do
+    vec1 <- freeze v1
+    vec2 <- freeze v2
+    return (vec1,vec2)
+
+runPaths' :: (Graph gr, Real b, Foldable f)
+    => Maybe Node -- Source node
+    -> Number b -- ^ Source node initialization value
+    -> Number b -- ^ Distances vector init values
+    -> (Number b -> Number b -> Bool) -- ^ Weight modifier
+    -> Int -- ^ Length of the top sort
+    -> f Node -- ^ Topological sorting of the graph in some foldable format
+    -> gr a b
+    -> IO (IOVector (Number b), IOVector (Maybe Node))
+runPaths' s v distV m n t g = do
+    (distances,predecessors) <- genVectors n distV Nothing
+    case s of 
+         Just source -> write distances source v
+         Nothing -> return ()
+    (newDistances,newPredecessors) <- pathsByComparison distances predecessors m t g
+    return (newDistances,newPredecessors)
+
+genVectors :: 
+    Int -- ^ Length of the vectors, ie, length of the graph or topsort
+    -> a
+    -> b
+    -> IO (IOVector a, IOVector b)
+genVectors n val1 val2 = do
+    distances <- new (n + 1)
```

Comment:
When you do, please benchmark them: for purposes of maintenance (and since the Vector solution has the potential of causing random breakages when someone uses it with a non-compact graph) it would be preferable to just have the map-based solution if the performance difference is minimal.